### PR TITLE
[WIP] Adding AMP support in benchmark infra

### PR DESCRIPTION
**Pull request by anijain2305**
Few things to notice

* Tests are failing accuracy left and right. Need to debug further why thats the case.
* AMP and optimizers are discussed in conjunction due to loss scaling. However, we have not yet tried optimizers with Dynamo. We will pick up the work after https://github.com/pytorch/torchdynamo/issues/328

cc @ngimel @Chillee

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: model_benchmark/common.py [hunk 1]
```diff
@@ -572,12 +572,72 @@ def cast_to_fp32(model, inputs):
...
+        self.use_amp = False
+        self.grad_scaler = DummyGradScaler()
+        self.autocast = NullContext
+        self._args = None
+
+    def setup_amp(self):
+        if self.args.amp and self.args.training:
+            assert self.args.devices == ["cuda"], "AMP is supported only for CUDA"
+            self.grad_scaler = torch.cuda.amp.GradScaler()
+            self.autocast = torch.cuda.amp.autocast
```
Comment (ngimel) [inline model_benchmark/common.py:591]: Nit: you can use generic autocast manager with `device="cuda"` https://pytorch.org/docs/stable/amp.html#torch.autocast
